make_travis_yml_2.hs

```diff
@@ -509,3 +649,253 @@ collToGhcVer cid = case simpleParse cid of
+        emptyLines <++ skipSpaces
+        char ','
+        emptyLines <++ skipSpaces
+
+    nonComma = emptyLines <++ satisfyP (>0) skipSpaces
+
+filePathGlob :: ReadP FilePathGlob
+filePathGlob = between (char '"') (char '"') (baseGlob True) <++ baseGlob False
+  where
+    nonEmpty = satisfyP (GlobDirTrailing/=)
+    baseGlob isQuoted = choice
+        [ FilePathGlob <$> filePathRoot <*> filePathGlobRel isQuoted
+        , FilePathGlob FilePathRelative <$> nonEmpty (filePathGlobRel isQuoted)
+        ]
+
+filePathRoot :: ReadP FilePathRoot
+filePathRoot = unixRoot <++ windowsRoot <++ homeDir
+  where
+    pathSep = char '\\' <++ char '/'
+    unixRoot = FilePathRoot . pure <$> char '/'
+    windowsRoot = do
+        drive <- satisfy $ \c -> isAsciiUpper c || isAsciiLower c
+        char ':'
+        sep <- pathSep
+        return $ FilePathRoot [drive,':',sep]
+    homeDir = FilePathHomeDir <$ char '~' <* pathSep
+
+filePathGlobRel :: Bool -> ReadP FilePathGlobRel
+filePathGlobRel b = nonEmptyFilePathGlobRel b <++ return GlobDirTrailing
+
+nonEmptyFilePathGlobRel :: Bool -> ReadP FilePathGlobRel
+nonEmptyFilePathGlobRel isQuoted = globDir <++ globFile
+  where
+    globDir = GlobDir <$> glob isQuoted <* (char '/' <++ char '\\')
+                      <*> filePathGlobRel isQuoted
+    globFile = GlobFile <$> glob isQuoted
+
+glob :: Bool -> ReadP Glob
+glob isQuoted = do
+    r <- globPiece isQuoted
+    (r:) <$> glob'
+  where
+    glob' :: ReadP Glob
+    glob' = ((:) <$> globPiece isQuoted <*> glob') <++ return []
+
+reservedChars :: [Char]
+reservedChars = "*{},\""
+
+globPiece :: Bool -> ReadP GlobPiece
+globPiece isQuoted = wildCard <++ union <++ literal
+  where
+    wildCard = WildCard <$ char '*'
+    union = between (char '{') (char '}') $
+        Union <$> sepBy (glob isQuoted) (char ',')
+
+    literal = Literal <$> many1 (safeChar <++ escapedChar)
+      where
+        notReserved c
+            | isQuoted = True
+            | otherwise = not (isSpace c)
+        safeChar = satisfy (\c -> c `notElem` "*{},/\\\"" && notReserved c)
+        escapedChar = char '\\' *> satisfy (`elem` reservedChars)
+
+expandGlobs :: FilePath -> [FilePathGlob] -> IO [FilePath]
+expandGlobs root globs = concat <$> mapM (expandGlob root) globs
+
+expandGlob :: FilePath -> FilePathGlob -> IO [FilePath]
+expandGlob _ g@(FilePathGlob FilePathHomeDir _) = do
+    putStrLnInfo $ "Ignoring home directory glob: " ++ prettyPathGlob g
+    return []
+expandGlob _ g@(FilePathGlob FilePathRoot{} _) = do
+    putStrLnInfo $ "Ignoring global directory glob: " ++ prettyPathGlob g
+    return []
+expandGlob root g@(FilePathGlob FilePathRelative relGlob) = do
+    result <- expandRelGlob root relGlob
+    when (null result) $
+        putStrLnErr $ "Illegal empty package glob: " ++ prettyPathGlob g
+    return result
+
+expandRelGlob :: FilePath -> FilePathGlobRel -> IO [FilePath]
+expandRelGlob root glob0 = go glob0 ""
+  where
+    go (GlobFile glob) dir = do
```

Comment:
is this globbing different from one we have for `extra-source-files`? Should we expose it from `Cabal` library (if not yet?)

Reply:
This globbing code is (mostly) ripped from the globbing code in cabal-install `Distribution.Client.Glob`. I don't think Cabal has equivalent globbing code?

Reply:
I think it's also more powerful/flexible than the extra-source-files globbing.